File: simulation/PID.py

import math
import numpy as np
import numpy.random as npr

from CC import *


def wma(avg, x):
    return (9*avg + x)/10

def clamp(x, minimum, maximum):
    if x < minimum:
        return minimum
    elif x > maximum:
        return maximum
    else:
        return x


class PID:
    def __init__(self, runInfo, bottleneckRate, baseRTT, coalesce):
        self.bottleneckRate = bottleneckRate
        self.baseRTT = baseRTT
        self.coalesce = coalesce

        self.minRate = min(self.bottleneckRate/128, runInfo.mss/self.baseRTT)

        self.srtt = runInfo.lastRTT
        self.srttLastTime = runInfo.time

        self.tau = max(4*baseRTT, runInfo.mss/self.bottleneckRate)
        self.targetRTT = baseRTT

        self.rateLastTime = runInfo.time
        self.targetLastTime = runInfo.time

        self.integ = self.minRate
        self.rate = self.minRate
        self.slowStart = True

        self.mu = self.minRate
        self.muDeliv = 0
        self.muTime = runInfo.time


    def pacingRate(self, runInfo):
        return self.rate

    def cwnd(self, runInfo):
        bdp = int(self.bottleneckRate*self.targetRTT/runInfo.mss)
        return max(4, 2*bdp)


    def ack(self, runInfo):
        self.minRate = min(self.bottleneckRate/128, runInfo.mss/self.srtt)
        self.tau = max(4*self.srtt, runInfo.mss/self.bottleneckRate)

        self.updateSRTT(runInfo)

        deliv = runInfo.delivered - self.muDeliv
        if runInfo.time > self.muTime + self.targetRTT and deliv > 0:
            est = deliv*runInfo.mss/(runInfo.time - self.muTime)

            if self.srtt > self.targetRTT:
                self.mu = est
            else:
                self.mu = max(self.mu, est)

            self.mu = clamp(self.mu, self.minRate, 2*self.bottleneckRate)

            self.muDeliv = runInfo.delivered
            self.muTime = runInfo.time

        if self.srtt > 2*self.targetRTT:
            self.slowStart = False

        self.update(runInfo)


    def loss(self, runInfo):
        self.slowStart = False
        self.update(runInfo)


    def update(self, runInfo):
        self.updateTargetRTT(runInfo)
        self.updateRate(runInfo)


    def updateRate(self, runInfo):
        dt = runInfo.time - self.rateLastTime

        err = self.targetRTT - runInfo.lastRTT

        kp = 2*self.mu/self.tau
        ki = self.mu/self.tau**2

        if self.slowStart:
            diffInteg = self.rate
            diffInteg *= min(1, dt/self.tau)
        else:
            diffInteg = ki*err
            diffInteg *= min(self.tau, dt)

        if diffInteg != 0:
            self.rateLastTime = runInfo.time

            self.integ += diffInteg
            self.rate = kp*err + self.integ

        self.integ = clamp(self.integ, self.minRate, 2*self.bottleneckRate)
        self.rate = clamp(self.rate, self.minRate, 2*self.bottleneckRate)


    def updateSRTT(self, runInfo):
        dt = runInfo.time - self.srttLastTime

        diffSRTT = runInfo.lastRTT - self.srtt
        diffSRTT *= min(1, dt/self.baseRTT)

        if diffSRTT != 0:
            self.srttLastTime = runInfo.time

            self.srtt += diffSRTT


    def updateTargetRTT(self, runInfo):
        dt = runInfo.time - self.targetLastTime

        nt = self.baseRTT
        nt += self.coalesce*runInfo.mss/self.rate
        nt += runInfo.hops*runInfo.mss/self.bottleneckRate

        diffTargetRTT = (nt - self.targetRTT)
        diffTargetRTT *= min(1, dt/self.tau)

        if diffTargetRTT != 0:
            self.targetLastTime = runInfo.time

            self.targetRTT += diffTargetRTT
            self.targetRTT = max(self.baseRTT, self.targetRTT)


    def getDebugInfo(self):
        return {'srtt': self.srtt,
                'tau': self.tau,
                'mu': self.mu,
                'integ': self.integ,
                'targetRTT': self.targetRTT,
                'slowStart': int(self.slowStart)}

File: src/pid.c

#include "pid.h"


void pid_update(struct pid_control *self, struct runtime_info *info);
void pid_update_rate(struct pid_control *self, struct runtime_info *info);
void pid_update_srtt(struct pid_control *self, struct runtime_info *info);
void pid_update_targetRTT(struct pid_control *self, struct runtime_info *info);


SNum pid_abs(SNum x) {
    if (x < 0)
        return -x;
    else
        return x;
}

SNum pid_min(SNum x, SNum y) {
    if (x < y)
        return x;
    else
        return y;
}

SNum pid_max(SNum x, SNum y) {
    if (x > y)
        return x;
    else
        return y;
}

SNum pid_clamp(SNum x, SNum min, SNum max) {
    if (x < min)
        return min;
    else if (x > max)
        return max;
    else
        return x;
}

SNum pid_div(SNum x, SNum y) {
    if (y == 0) {
        #ifdef PID_KERNEL_MODE
        printk("pid_cc: Division by zero, %lld/%lld\n", x, y);
        #endif

        return x;
    } else {
        return x/y;
    }
}

SNum pid_sqr(SNum x) {
    return x*x;
}

SNum pid_wma(SNum avg, SNum x) {
    return (9*avg + x)/10;
}



void pid_init(struct pid_control *self, const struct pid_config *cfg, struct runtime_info *info) {
    self->cfg = *cfg;
    pid_reset(self, info);
}


void pid_release(struct pid_control *self) {
}


void pid_reset(struct pid_control *self, struct runtime_info *info) {
    self->minRate = pid_min(self->cfg.bottleneckRate/128,
                            pid_div(info->mss*US_PER_SEC, self->cfg.baseRTT));
    self->minRate = pid_max(minDelta, self->minRate);

    self->srtt = info->lastRTT;
    self->srttLastTime = info->time;

    self->tau = pid_max(4*self->cfg.baseRTT,
                        pid_div(info->mss*US_PER_SEC, self->cfg.bottleneckRate));

    self->targetRTT = self->cfg.baseRTT
        + pid_div(info->hops*info->mss, self->minRate);

    self->rateLastTime = info->time;
    self->targetLastTime = info->time;

    self->integ = self->minRate;
    self->rate = self->minRate;
    self->slowStart = true;

    self->mu = self->cfg.bottleneckRate;
    self->muDeliv = 0;
    self->muTime = info->time;
}



SNum pid_pacing_rate(struct pid_control *self, struct runtime_info *info) {
    return self->rate;
}


SNum pid_cwnd(struct pid_control *self, struct runtime_info *info) {
    SNum bdp;

    if (info->mss < 1)
        bdp = self->cfg.bottleneckRate*self->targetRTT/US_PER_SEC;
    else
        bdp = pid_div(self->cfg.bottleneckRate*self->targetRTT, info->mss*US_PER_SEC);

    return pid_max(4, bdp);
}


void pid_on_ack(struct pid_control *self, struct runtime_info *info) {
    SNum deliv, est;

    self->minRate = pid_min(self->cfg.bottleneckRate/128,
                            pid_div(info->mss*US_PER_SEC, self->srtt));
    self->minRate = pid_max(minDelta, self->minRate);

    self->tau = pid_max(4*self->targetRTT,
                        pid_div(info->mss*US_PER_SEC, self->cfg.bottleneckRate));

    pid_update_srtt(self, info);


    deliv = info->delivered - self->muDeliv;
    if (info->time > self->muTime + self->targetRTT && deliv > 0) {
        est = pid_div(deliv*info->mss*US_PER_SEC, info->time - self->muTime);

        if (self->srtt > self->targetRTT)
            self->mu = est;
        else
            self->mu = pid_max(self->mu, est);

        self->mu = pid_clamp(self->mu, self->minRate, 2*self->cfg.bottleneckRate);

        self->muDeliv = info->delivered;
        self->muTime = info->time;
    }


    pid_update(self, info);
}


void pid_on_loss(struct pid_control *self, struct runtime_info *info) {
    self->slowStart = false;
    pid_update(self, info);
}


void pid_update(struct pid_control *self, struct runtime_info *info) {
    pid_update_targetRTT(self, info);
    pid_update_rate(self, info);
}


void pid_update_rate(struct pid_control *self, struct runtime_info *info) {
    SNum dt, err;
    SNum kpNum, kpDen, kiNum, kiDen;
    SNum diffInteg;

    dt = info->time - self->rateLastTime;

    if (dt < 1)
        return;

    err = self->targetRTT - info->lastRTT;

    kpNum = 2*self->mu;
    kpDen = self->tau;

    kiNum = self->mu;
    kiDen = pid_sqr(self->tau);

    if (self->slowStart) {
        diffInteg = pid_div(self->rate*pid_min(self->tau, dt), self->tau);
    } else {
        diffInteg = pid_div(kiNum*err*pid_min(self->tau, dt), kiDen);
    }

    if (diffInteg != 0) {
        self->rateLastTime = info->time;

        self->integ += diffInteg;
        self->rate = pid_div(kpNum*err, kpDen) + self->integ;
    }

    self->integ = pid_clamp(self->integ, self->minRate, 2*self->cfg.bottleneckRate);
    self->rate = pid_clamp(self->rate, self->minRate, 2*self->cfg.bottleneckRate);
}


void pid_update_srtt(struct pid_control *self, struct runtime_info *info) {
    SNum dt;
    SNum diffSRTT;

    dt = info->time - self->srttLastTime;

    if (dt < 1)
        return;

    diffSRTT = info->lastRTT - self->srtt;
    diffSRTT = pid_div(diffSRTT*pid_min(self->cfg.baseRTT, dt), self->cfg.baseRTT);

    if (diffSRTT != 0) {
        self->srttLastTime = info->time;
        self->srtt += diffSRTT;

        self->srtt = pid_max(minDelta, self->srtt);
    }
}


void pid_update_targetRTT(struct pid_control *self, struct runtime_info *info) {
    SNum dt, nt;
    SNum diffTargetRTT;

    dt = info->time - self->targetLastTime;

    if (dt < 1)
        return;

    nt = self->cfg.baseRTT;
    nt += pid_div(self->cfg.coalesce*info->mss*US_PER_SEC, self->rate);
    nt += pid_div(info->hops*info->mss*US_PER_SEC, self->cfg.bottleneckRate);

    diffTargetRTT = pid_div((nt - self->targetRTT)*pid_min(self->tau, dt), self->tau);


    if (diffTargetRTT != 0) {
        self->targetLastTime = info->time;

        self->targetRTT += diffTargetRTT;
    }

    self->targetRTT = pid_max(self->cfg.baseRTT, self->targetRTT);
}
